text/0083-rawkv-cross-cluster-replication.md

```diff
@@ -67,9 +67,11 @@ Physical part of HLC is acquired from TSO of [PD](https://github.com/tikv/pd).
 
 TSO is a global monotonically increasing timestamp, which help the generation of timestamp be independent to local clock of machine, and be immune to issues such as reverse between machine reboot.
 
-Physical part is refreshed by repeatedly acquiring TSO in a period of `500ms`, to keep it being closed the real world time. And it can tolerate fault of TSO no longer than `30s`, to keep time-related metrics such as RPO reasonable.
+Physical part is refreshed by repeatedly acquiring TSO in a period of `500ms`, to keep it being closed the real world time.
 
-Besides, on startup, physical part must be initialized by a successful TSO.
+On startup, physical part must be initialized by a successful TSO, to ensure that HLC is larger than last running. RawKV write operations will fail if the initialization has not been completed. (TiKV also can't startup when PD is out of service).
+
+After that, HLC can tolerate fault of TSO. RawKV operations are normal, but time-related metrics such as RPO is not reasonable. Meanwhile, as TiKV-CDC being highly dependent to PD (for cluster and task management, checkpoint, etc.), the replication will not be working.
```

Comment:
What exactly is the "After **that**"? From the context, it is the TSO failure on startup but I think it should be TSO failure after startup. And I think the time-related metrics may not be affected unless the clock shift is crazy enough...

Reply:
It is "After successful initialization". I'm sorry for the not clear enough.

We do not advance HLC by local clock when TSO is offline, which will cause the time-related metrics problem.

Using local clock on fault of TSO need some other logics to ensure causality, such as saving maximum HLC on storage to avoid timestamp reverse between startup.

I do not include the "Falling Back to Local Clock" in this proposal for simplicity, as the PD offline is rare, and when it was offline it would cause more severe issues than the problem of time-related metrics.

Reply:
> Using local clock on fault of TSO need some other logics to ensure causality, such as saving maximum HLC on storage to avoid timestamp reverse between startup.

If I understand it right, the `RegionMaxTs` and `StoreMaxTs` should contain both physical time and logical time, thus we always know what is the latest physical time and therefore safe to use the local clock.

Reply:
Yes it is safe during running.
But unsafe will happen on restart. If the local clock shift backward between TiKV stop & restart, it's possible to get a smaller timestamp than last running.

Reply:
CockroachDB has a description on this restart issue [here](https://github.com/cockroachdb/cockroach/blob/d5ee674ec410f384df9ffb6f737e9b4e2c922589/pkg/util/hlc/doc.go#L152-L160). We can refer to it.

Reply:
> Yes it is safe during running.

Yes, that's my point. As long as the TiKV doesn't restart, the metrics of RPO and write operation should not affect by PD failure.

Reply:
it would be better if you can explain the reason with an example. For example:

In a 3 AZ deployment:
- PD leader is located in AZ1
- TiKV region leaders are uniformly distributed in AZ1 - AZ3
- There is a partial network partition that happens: AZ1 can not communicate with AZ2, but they can both communicate with AZ3, TiKV Client (for example, client-java) can connect to all the three AZs

Would the write requests to AZ2 TiKV Region be blocked because AZ2 can not sync physical timestamp with PD leader in AZ1?

Reply:
I improve the words here. PTAL~